### PR TITLE
[#403] Configurable defaults in HonoClient

### PR DESCRIPTION
**Pull request by pellmann**

### CODE REVIEW

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 1]
```diff
@@ -145,8 +145,8 @@ public final void setRequestTimeout(final long timoutMillis) {
      * @return A future indicating the outcome. The future will succeed if the links
      *         have been created.
      */
-    protected final Future<Void> createLinks(final ProtonConnection con) {
-        return createLinks(con, null, null);
+    protected final Future<Void> createLinks(final ProtonConnection con, final int receiverPrefetchCredits, final long waitForInitialCredits) {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: `@param` declaration missing

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 2]
```diff
@@ -159,11 +159,12 @@ public final void setRequestTimeout(final long timoutMillis) {
      * @return A future indicating the outcome. The future will succeed if the links
      *         have been created.
      */
-    protected final Future<Void> createLinks(final ProtonConnection con, final Handler<String> senderCloseHook, final Handler<String> receiverCloseHook) {
+    protected final Future<Void> createLinks(final ProtonConnection con, final int receiverPrefetchCredits, final long waitForInitialCredits,
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: `@param` declaration missing

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 3]
```diff
@@ -159,11 +159,12 @@ public final void setRequestTimeout(final long timoutMillis) {
...
+                                             final Handler<String> senderCloseHook, final Handler<String> receiverCloseHook) {
         Future<Void> result = Future.future();
-        createReceiver(con, replyToAddress, receiverCloseHook).compose(recv -> {
+        createReceiver(con, replyToAddress, receiverPrefetchCredits, receiverCloseHook).compose(recv -> {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: what about sanity checks for credits (> 0)?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java [hunk 4]
```diff
@@ -156,6 +151,7 @@ protected static final void setApplicationProperties(final Message msg, final Ma
             final ProtonConnection con,
             final String targetAddress,
             final ProtonQoS qos,
+            final long waitForInitialCredits,
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: `@param` declaration missing

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java [hunk 5]
```diff
@@ -166,14 +162,14 @@ protected static final void setApplicationProperties(final Message msg, final Ma
             sender.openHandler(senderOpen -> {
                 if (senderOpen.succeeded()) {
                     LOG.info("sender open [{}] sendQueueFull [{}]", sender.getRemoteTarget(), sender.sendQueueFull());
-                    // wait on credits a little time, if not already given TODO: configurable waiting time
+                    // wait on credits a little time, if not already given
                     if (sender.sendQueueFull()) {
-                        ctx.owner().setTimer(10, timerID -> {
-                            LOG.info("waited 10ms on credits [{}]", sender.getCredit());
-                            result.complete(senderOpen.result());
+                        ctx.owner().setTimer(waitForInitialCredits, timerID -> {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: sanity check? `waitForInitialCredits` >= 0? 0 = wait forever?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java [hunk 6]
```diff
@@ -217,14 +212,16 @@ protected static final void setApplicationProperties(final Message msg, final Ma
             final ProtonConnection con,
             final String sourceAddress,
             final ProtonQoS qos,
+            final int prefetchCredits,
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: `@param` declaration missing

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java [hunk 7]
```diff
@@ -217,14 +212,16 @@ protected static final void setApplicationProperties(final Message msg, final Ma
...
             final ProtonMessageHandler messageHandler,
             final Handler<String> closeHook) {
 
         Future<ProtonReceiver> result = Future.future();
         ctx.runOnContext(go -> {
             final ProtonReceiver receiver = con.createReceiver(sourceAddress);
             receiver.setAutoAccept(true);
-            receiver.setPrefetch(DEFAULT_SENDER_CREDITS);
+            receiver.setQoS(qos);
+            receiver.setPrefetch(prefetchCredits);
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: sanity checks? prefetchCredits > 0

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java [hunk 8]
```diff
@@ -159,11 +159,12 @@ public final void setRequestTimeout(final long timoutMillis) {
...
             this.receiver = recv;
-            return createSender(con, targetAddress, senderCloseHook);
+            return createSender(con, targetAddress, waitForInitialCredits, senderCloseHook);
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java]: sanity checks?

File: client/src/main/java/org/eclipse/hono/client/impl/EventSenderImpl.java [hunk 9]
```diff
@@ -93,7 +95,7 @@ public static void create(
         Objects.requireNonNull(creationHandler);
 
         final String targetAddress = getTargetAddress(tenantId, deviceId);
-        createSender(context, con, targetAddress, ProtonQoS.AT_LEAST_ONCE, closeHook).setHandler(created -> {
+        createSender(context, con, targetAddress, ProtonQoS.AT_LEAST_ONCE, waitForInitialCredits, closeHook).setHandler(created -> {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/EventSenderImpl.java]: sanity checks?

File: client/src/main/java/org/eclipse/hono/client/impl/CredentialsClientImpl.java [hunk 10]
```diff
@@ -85,12 +87,14 @@ public static void create(
...
             final String tenantId,
+            final int receiverPrefetchCredits,
+            final long waitForInitialCredits,
             final Handler<String> senderCloseHook,
             final Handler<String> receiverCloseHook,
             final Handler<AsyncResult<CredentialsClient>> creationHandler) {
 
         final CredentialsClientImpl client = new CredentialsClientImpl(context, tenantId);
-        client.createLinks(con, senderCloseHook, receiverCloseHook).setHandler(s -> {
+        client.createLinks(con, receiverPrefetchCredits, waitForInitialCredits, senderCloseHook, receiverCloseHook).setHandler(s -> {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/CredentialsClientImpl.java]: sanity checks?

File: client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java [hunk 11]
```diff
@@ -66,13 +85,14 @@ public HonoClientImpl(final Vertx vertx, final ConnectionFactory connectionFacto
             this.vertx = Vertx.vertx();
         }
         this.connectionFactory = connectionFactory;
+        this.clientConfigProperties = new ClientConfigProperties();
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java]: this constructor should simply invoke
`this(vertx, connectionFactory, new ClientConfigProperties())`

File: client/src/main/java/org/eclipse/hono/client/impl/RegistrationClientImpl.java [hunk 12]
```diff
@@ -74,12 +76,14 @@ public static void create(
...
             final String tenantId,
+            final int receiverPrefetchCredits,
+            final long waitForInitialCredits,
             final Handler<String> senderCloseHook,
             final Handler<String> receiverCloseHook,
             final Handler<AsyncResult<RegistrationClient>> creationHandler) {
 
         final RegistrationClientImpl client = new RegistrationClientImpl(context, tenantId);
-        client.createLinks(con, senderCloseHook, receiverCloseHook).setHandler(s -> {
+        client.createLinks(con, receiverPrefetchCredits, waitForInitialCredits, senderCloseHook, receiverCloseHook).setHandler(s -> {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/RegistrationClientImpl.java]: sanity checks?

File: client/src/main/java/org/eclipse/hono/client/impl/TelemetrySenderImpl.java [hunk 13]
```diff
@@ -91,7 +93,7 @@ public static void create(
         Objects.requireNonNull(creationHandler);
 
         final String targetAddress = getTargetAddress(tenantId, deviceId);
-        createSender(context, con, targetAddress, ProtonQoS.AT_MOST_ONCE, closeHook).setHandler(created -> {
+        createSender(context, con, targetAddress, ProtonQoS.AT_MOST_ONCE, waitForInitialCredits, closeHook).setHandler(created -> {
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/TelemetrySenderImpl.java]: sanity checks?

File: core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java [hunk 14]
```diff
@@ -157,4 +160,58 @@ public final String getAmqpHostname() {
     public final void setAmqpHostname(String amqpHostname) {
         this.amqpHostname = amqpHostname;
     }
+
+    /**
+     * Gets millis to wait for a receivers AMQP <em>flow</em> frame after sender link creation.
+     *
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java]: document default value

File: core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java [hunk 15]
```diff
@@ -157,4 +160,58 @@ public final String getAmqpHostname() {
...
+     * @param waitMillisForCredits The millis to set
+     */
+    public void setWaitMillisForInitialCredits(final long waitMillisForCredits) {
+        this.waitMillisForCredits = waitMillisForCredits;
+    }
+
+    /**
+     * Gets the number of initial credits, that will be given from a receiver to a sender at link creation.
+     *
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java]: document default value

File: core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java [hunk 16]
```diff
@@ -157,4 +160,58 @@ public final String getAmqpHostname() {
...
+     * @param initialCredits The initial credits to set.
+     */
+    public void setInitialCredits(final int initialCredits) {
+        this.initialCredits = initialCredits;
+    }
+
+    /**
+     * Gets the timeout in millis between a request and the awaited response in a request/response style communication
+     *
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java]: document default value

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java (hunk 5 above)
Comment (pellmann) [reply client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: The timer from vertx already does a check, that it need to be >= 1.

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java (hunk 7 above)
Comment (pellmann) [reply client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: Vertx proton checks for >= 0. 0 might be useful for own flow control?

**Comment by pellmann** [comment]
In regards to the sanity checks it is more a general question if we should also add checks (maybe the same, as the underlying framework) to give a more specific error. For me it might a bit overkill? Of course, the API of the framework might not define this, and we would be more on the sure side, if we do it ourself .. but still I do not think, this is likely to change in the frameworks. WDYT?

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java (hunk 5 above)
Comment (sophokles73) [reply client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: and what happens if it isn't? is an exception thrown? if so, then it should be declared in the JavaDocs ...

File: client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java (hunk 7 above)
Comment (sophokles73) [reply client/src/main/java/org/eclipse/hono/client/impl/AbstractHonoClient.java]: yes, still: what happens if < 0? if an exception is thrown, it should be declared in the JavaDocs

**Comment by sophokles73** [comment]
If we do not *wrap* exceptions from the underlying frameworks then we need to at least declare them being thrown in our API doc.

**Comment by pellmann** [comment]
I added the possible exceptions to the Javadocs and added the missing params - good eye on that. :+1:

**Pull request merged**